src/server/pps/server/pipeline_controller.go

```diff
@@ -188,48 +190,34 @@ func (pc *pipelineController) step(timestamp time.Time) (retErr error) {
 		tracing.TagAnySpan(span, "pollpipelines-event", "true")
 	}
 	defer func() {
-		tracing.FinishAnySpan(span, "err", retErr)
+		tracing.FinishAnySpan(span, "err", finalErr)
```

Comment:
If you're going to read `finalErr` in a defer statement, I think it might make sense to make `finalErr` a return value and return it, even if the caller doesn't read it. I think that the risk of using a local variable in a defer statement is that local variables are easy to accidentally shadow, but if it's a return variable, then anywhere that calls `return x, err` will work.